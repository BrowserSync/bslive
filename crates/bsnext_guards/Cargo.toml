[package]
name = "bsnext_guards"
version = "0.3.1"
edition = "2021"

[dependencies]
urlpattern = { version = "0.3.0" }
serde = { workspace = true }
tracing = { workspace = true }
axum = { workspace = true }
http = { workspace = true }
