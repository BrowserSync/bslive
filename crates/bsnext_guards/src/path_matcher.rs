use crate::root_path::{RootPath, RootPathError};
use http::Uri;
use std::str::FromStr;
use tracing::trace_span;
use urlpattern::UrlPatternInit;
use urlpattern::UrlPatternMatchInput;
use urlpattern::{UrlPattern, UrlPatternOptions};

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
#[serde(untagged)]
pub enum PathMatcher {
    Str(#[serde(with = "crate::root_path")] RootPath),
    Def(PathMatcherDef),
}

impl FromStr for PathMatcher {
    type Err = RootPathError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(PathMatcher::Str(RootPath::from_str(s)?))
    }
}

impl PathMatcher {
    pub fn pathname(str: impl Into<String>) -> Self {
        Self::Def(PathMatcherDef {
            pathname: Some(str.into()),
            search: None,
        })
    }
    pub fn query(str: impl Into<String>) -> Self {
        Self::Def(PathMatcherDef {
            pathname: None,
            search: Some(str.into()),
        })
    }
}

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
pub struct PathMatcherDef {
    pub(crate) pathname: Option<String>,
    pub(crate) search: Option<String>,
}

impl PathMatcher {
    pub fn test_uri(&self, uri: &Uri) -> bool {
        let span = trace_span!("test_uri", uri = ?uri, path_matcher = ?self);
        let _g = span.enter();
        let Some(path_and_query) = uri.path_and_query() else {
            tracing::error!("how is this possible?");
            return false;
        };

        let path = path_and_query.path();
        let seary = path_and_query.query();

        let incoming = UrlPatternInit {
            pathname: Some(path.into()),
            search: seary.map(ToOwned::to_owned),
            ..Default::default()
        };

        tracing::trace!(?incoming);

        // convert the config into UrlPatternInit
        // example: /style.css
        let matching_options: UrlPatternInit = match self {
            PathMatcher::Str(str) => {
                if let Ok(uri) = &Uri::from_str(str.as_str()) {
                    if let Some(pq) = uri.path_and_query() {
                        let path = pq.path();
                        let query = pq.query();
                        UrlPatternInit {
                            pathname: Some(path.into()),
                            search: query.map(ToOwned::to_owned),
                            ..Default::default()
                        }
                    } else {
                        tracing::trace!(?str, "path and query was missing");
                        Default::default()
                    }
                } else {
                    tracing::trace!(?str, "Uri::from_str failed");
                    Default::default()
                }
            }
            PathMatcher::Def(PathMatcherDef { pathname, search }) => UrlPatternInit {
                pathname: pathname.to_owned(),
                search: search.to_owned(),
                ..Default::default()
            },
        };
        let opts = UrlPatternOptions::default();
        tracing::trace!(?opts);

        let Ok(pattern) = <UrlPattern>::parse(matching_options, opts) else {
            tracing::error!("could not parse the input");
            return false;
        };
        tracing::trace!(?pattern);

        match pattern.test(UrlPatternMatchInput::Init(incoming)) {
            Ok(true) => {
                tracing::trace!("matched!");
                true
            }
            Ok(false) => {
                tracing::trace!("not matched!");
                false
            }
            Err(e) => {
                tracing::error!("could not match {:?}", e);
                false
            }
        }
    }
}

#[cfg(test)]
mod test {
    use crate::path_matcher::PathMatcher;
    use http::Uri;
    use std::str::FromStr;

    #[test]
    fn test_url_pattern_pathname() {
        let pm = PathMatcher::pathname("/");
        assert_eq!(pm.test_uri(&Uri::from_static("/")), true);
        let pm = PathMatcher::pathname("/*.css");
        assert_eq!(pm.test_uri(&Uri::from_static("/style.css")), true);
        let pm = PathMatcher::pathname("/here/*.css");
        assert_eq!(pm.test_uri(&Uri::from_static("/style.css")), false);
        let pm = PathMatcher::pathname("/**/*.css");
        assert_eq!(pm.test_uri(&Uri::from_static("/style.css")), true);
        let pm = PathMatcher::pathname("/**/*.css");
        assert_eq!(
            pm.test_uri(&Uri::from_static("/a/b/c/--oopasxstyle.css")),
            true
        );
        assert_eq!(
            pm.test_uri(&Uri::from_static("/a/b/c/--oopasxstyle.html")),
            false
        );
    }
    #[test]
    fn test_url_pattern_query() -> anyhow::Result<()> {
        let pm = PathMatcher::from_str("/?abc=true")?;
        assert_eq!(pm.test_uri(&Uri::from_static("/")), false);
        assert_eq!(pm.test_uri(&Uri::from_static("/?def=true")), false);
        assert_eq!(pm.test_uri(&Uri::from_static("/?abc=true")), true);
        assert_eq!(pm.test_uri(&Uri::from_static("/?abc=")), false);

        let pm2 = PathMatcher::from_str("/**/*?delayms")?;
        assert_eq!(pm2.test_uri(&Uri::from_static("/?delayms")), true);

        let pm2 = PathMatcher::query("?*a*b*c*foo=bar");
        assert_eq!(
            pm2.test_uri(&Uri::from_static("/?delay.ms=2000&a-b-c-foo=bar")),
            true
        );
        Ok(())
    }
    #[test]
    fn test_url_pattern_str() -> anyhow::Result<()> {
        let pm = PathMatcher::from_str("/")?;
        assert_eq!(pm.test_uri(&Uri::from_static("/")), true);
        let pm = PathMatcher::from_str("/*.css")?;
        assert_eq!(pm.test_uri(&Uri::from_static("/style.css")), true);
        let pm = PathMatcher::from_str("/here/*.css")?;
        assert_eq!(pm.test_uri(&Uri::from_static("/style.css")), false);
        let pm = PathMatcher::from_str("/**/*.css")?;
        assert_eq!(pm.test_uri(&Uri::from_static("/style.css")), true);
        let pm = PathMatcher::from_str("/**/*.css")?;
        assert_eq!(
            pm.test_uri(&Uri::from_static("/a/b/c/--oopasxstyle.css")),
            true
        );
        assert_eq!(
            pm.test_uri(&Uri::from_static("/a/b/c/--oopasxstyle.html")),
            false
        );
        Ok(())
    }
}
