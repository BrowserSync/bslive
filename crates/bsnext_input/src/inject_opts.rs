#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
#[serde(untagged)]
pub enum InjectOpts {
    Bool(bool),
    Items(Vec<String>),
}

impl InjectOpts {
    pub fn injections(&self) -> Vec<String> {
        match self {
            InjectOpts::Bool(true) => {
                vec!["bslive:connector".to_string()]
            }
            InjectOpts::Bool(false) => {
                vec![]
            }
            InjectOpts::Items(_) => todo!("implement injection list"),
        }
    }
}

impl Default for InjectOpts {
    fn default() -> Self {
        Self::Bool(true)
    }
}
