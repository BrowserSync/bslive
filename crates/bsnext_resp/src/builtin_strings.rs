use crate::connector::Connector;
use crate::debug::Debug;
use crate::injector_guard::ByteReplacer;
use crate::js_connector::JsConnector;
use axum::extract::Request;
use bsnext_guards::route_guard::RouteGuard;
use http::Response;

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
pub struct BuiltinStringDef {
    pub name: BuiltinStrings,
}
#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
pub enum BuiltinStrings {
    #[serde(rename = "bslive:connector")]
    Connector,
    #[serde(rename = "bslive:js-connector")]
    JsConnector,
    #[serde(rename = "bslive:debug")]
    Debug,
}

impl RouteGuard for BuiltinStringDef {
    fn accept_req(&self, req: &Request) -> bool {
        match self.name {
            BuiltinStrings::Connector => Connector.accept_req(req),
            BuiltinStrings::Debug => Debug.accept_req(req),
            BuiltinStrings::JsConnector => JsConnector.accept_req(req),
        }
    }

    fn accept_res<T>(&self, res: &Response<T>) -> bool {
        match self.name {
            BuiltinStrings::Connector => Connector.accept_res(res),
            BuiltinStrings::Debug => Debug.accept_res(res),
            BuiltinStrings::JsConnector => JsConnector.accept_res(res),
        }
    }
}

impl ByteReplacer for BuiltinStringDef {
    fn apply(&self, body: &'_ str) -> Option<String> {
        match self.name {
            BuiltinStrings::Connector => Connector.apply(body),
            BuiltinStrings::Debug => Debug.apply(body),
            BuiltinStrings::JsConnector => JsConnector.apply(body),
        }
    }
}
