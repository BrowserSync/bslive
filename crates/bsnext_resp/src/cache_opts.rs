use http::header::{CACHE_CONTROL, EXPIRES, PRAGMA};
use http::{HeaderName, HeaderValue};

#[derive(Debug, Default, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
pub enum CacheOpts {
    /// Try to prevent browsers from caching the responses. This is the default behaviour
    #[serde(rename = "prevent")]
    #[default]
    Prevent,
    /// Do not try to prevent browsers from caching the responses.
    #[serde(rename = "default")]
    Default,
}

impl CacheOpts {
    pub fn as_headers(&self) -> Vec<(HeaderName, HeaderValue)> {
        match self {
            CacheOpts::Prevent => {
                vec![
                    (
                        CACHE_CONTROL,
                        HeaderValue::from_static("no-store, no-cache, must-revalidate"),
                    ),
                    (PRAGMA, HeaderValue::from_static("no-cache")),
                    (EXPIRES, HeaderValue::from_static("0")),
                ]
            }
            CacheOpts::Default => {
                vec![]
            }
        }
    }
}
