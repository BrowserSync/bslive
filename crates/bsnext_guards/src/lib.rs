use crate::path_matcher::PathMatcher;
use axum::extract::Request;
use axum::middleware::Next;
use axum::response::IntoResponse;
use http::Uri;

pub mod path_matcher;
pub mod root_path;
pub mod route_guard;

#[derive(Debug, Default, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
#[serde(untagged)]
pub enum MatcherList {
    #[default]
    None,
    Item(PathMatcher),
    Items(Vec<PathMatcher>),
}

impl MatcherList {
    pub fn test_uri(&self, uri: &Uri) -> bool {
        match self {
            MatcherList::None => true,
            MatcherList::Item(matcher) => matcher.test_uri(uri),
            MatcherList::Items(matchers) => matchers.iter().any(|m| m.test_uri(uri)),
        }
    }
}

#[derive(Debug, Clone)]
pub struct OuterUri(pub Uri);
pub async fn uri_extension(uri: Uri, mut req: Request, next: Next) -> impl IntoResponse {
    req.extensions_mut().insert(OuterUri(uri));
    next.run(req).await
}
