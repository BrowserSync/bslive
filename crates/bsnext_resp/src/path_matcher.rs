use urlpattern::UrlPattern;
use urlpattern::UrlPatternInit;
use urlpattern::UrlPatternMatchInput;

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
#[serde(untagged)]
pub enum PathMatcher {
    Str(String),
    Def(PathMatcherDef),
}

impl PathMatcher {
    pub fn pathname(str: impl Into<String>) -> Self {
        Self::Def(PathMatcherDef {
            pathname: Some(str.into()),
        })
    }
}

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
pub struct PathMatcherDef {
    pub pathname: Option<String>,
}

impl PathMatcher {
    pub fn test(&self, uri: &str) -> bool {
        let incoming = UrlPatternInit {
            pathname: Some(uri.to_owned()),
            ..Default::default()
        };

        let to_pathname = match self {
            PathMatcher::Str(str) => str.as_str(),
            PathMatcher::Def(PathMatcherDef {
                pathname: Some(str),
            }) => str.as_str(),
            PathMatcher::Def(PathMatcherDef { pathname: None }) => {
                unreachable!("how can this occur?")
            }
        };
        tracing::trace!(?to_pathname, ?uri, "PathMatcher::Str");
        let init = UrlPatternInit {
            pathname: Some(to_pathname.to_owned()),
            ..Default::default()
        };
        let Ok(pattern) = <UrlPattern>::parse(init) else {
            tracing::error!(?to_pathname, "could not parse the input");
            return false;
        };
        match pattern.test(UrlPatternMatchInput::Init(incoming)) {
            Ok(true) => {
                tracing::trace!("matched!");
                true
            }
            Ok(false) => {
                tracing::trace!("not matched!");
                false
            }
            Err(e) => {
                tracing::error!("could not match {:?}", e);
                false
            }
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::inject_addition::{AdditionPosition, InjectAddition};
    use crate::inject_opts::{InjectOpts, Injection, InjectionItem, MatcherList};

    #[test]
    fn test_path_matchers() {
        #[derive(Debug, serde::Deserialize)]
        struct A {
            inject: InjectOpts,
        }
        let input = r#"
inject:
    append: lol
    only:
      - /*.css
      - pathname: /*.css
"#;
        let expected = A {
            inject: InjectOpts::Item(InjectionItem {
                inner: Injection::Addition(InjectAddition {
                    addition_position: AdditionPosition::Append("lol".to_string()),
                }),
                only: Some(MatcherList::Items(vec![
                    PathMatcher::Str("/*.css".to_string()),
                    PathMatcher::Def(PathMatcherDef {
                        pathname: Some("/*.css".to_string()),
                    }),
                ])),
            }),
        };
        let actual: Result<A, _> = serde_yaml::from_str(input);
        assert_eq!(actual.unwrap().inject, expected.inject);
    }

    #[test]
    fn test_path_matcher_single() {
        #[derive(Debug, serde::Deserialize)]
        struct A {
            inject: InjectOpts,
        }
        let input = r#"
    inject:
        append: lol
        only: /*.css
    "#;
        let expected = A {
            inject: InjectOpts::Item(InjectionItem {
                inner: Injection::Addition(InjectAddition {
                    addition_position: AdditionPosition::Append("lol".to_string()),
                }),
                only: Some(MatcherList::Item(PathMatcher::Str("/*.css".to_string()))),
            }),
        };
        let actual: Result<A, _> = serde_yaml::from_str(input);
        assert_eq!(actual.unwrap().inject, expected.inject);
    }

    #[test]
    fn test_url_pattern() {
        let pm = PathMatcher::pathname("/");
        assert_eq!(pm.test("/"), true);
        let pm = PathMatcher::pathname("/*.css");
        assert_eq!(pm.test("/style.css"), true);
        let pm = PathMatcher::pathname("/here/*.css");
        assert_eq!(pm.test("/style.css"), false);
        let pm = PathMatcher::pathname("/**/*.css");
        assert_eq!(pm.test("/style.css"), true);
        let pm = PathMatcher::pathname("/**/*.css");
        assert_eq!(pm.test("/a/b/c/--oopasxstyle.css"), true);
        assert_eq!(pm.test("/a/b/c/--oopasxstyle.html"), false);
    }
}
