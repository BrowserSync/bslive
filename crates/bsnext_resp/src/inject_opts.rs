use crate::builtin_strings::BuiltinStrings;
use crate::inject_addition::InjectAddition;
use crate::inject_replacement::InjectReplacement;
use crate::injector_guard::{ByteReplacer, InjectorGuard};
use axum::extract::Request;
use http::Response;

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
#[serde(untagged)]
pub enum InjectOpts {
    Bool(bool),
    Item(Injection),
    Items(Vec<Injection>),
}

impl InjectOpts {
    pub fn injections(&self) -> Vec<Injection> {
        match self {
            InjectOpts::Bool(true) => {
                vec![Injection::BsLive(BuiltinStrings::Connector)]
            }
            InjectOpts::Bool(false) => {
                vec![]
            }
            InjectOpts::Items(items) if items.is_empty() => vec![],
            // todo: is this too expensive?
            InjectOpts::Items(items) => items.to_owned(),
            InjectOpts::Item(item) => vec![item.to_owned()],
        }
    }
}

impl Default for InjectOpts {
    fn default() -> Self {
        Self::Bool(true)
    }
}

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
#[serde(untagged)]
pub enum Injection {
    BsLive(BuiltinStrings),
    UnknownNamed(String),
    Replacement(InjectReplacement),
    Addition(InjectAddition),
}

impl InjectorGuard for Injection {
    fn accept_req(&self, req: &Request) -> bool {
        match self {
            Injection::BsLive(built_ins) => built_ins.accept_req(req),
            Injection::UnknownNamed(_) => todo!("accept_req Injection::UnknownNamed"),
            Injection::Replacement(def) => def.accept_req(req),
            Injection::Addition(add) => add.accept_req(req),
        }
    }

    fn accept_res<T>(&self, res: &Response<T>) -> bool {
        match self {
            Injection::BsLive(built_ins) => built_ins.accept_res(res),
            Injection::UnknownNamed(_) => todo!("accept_res Injection::UnknownNamed"),
            Injection::Replacement(def) => def.accept_res(res),
            Injection::Addition(add) => add.accept_res(res),
        }
    }
}
impl ByteReplacer for Injection {
    fn apply(&self, body: &'_ str) -> Option<String> {
        match self {
            Injection::BsLive(strs) => strs.apply(body),
            Injection::UnknownNamed(_) => todo!("Injection::UnknownNamed"),
            Injection::Replacement(def) => def.apply(body),
            Injection::Addition(add) => add.apply(body),
        }
    }
}
