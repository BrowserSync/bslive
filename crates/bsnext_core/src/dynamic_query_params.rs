use axum::extract::{Query, Request};
use axum::middleware::Next;
use axum::response::IntoResponse;
use bsnext_guards::path_matcher::PathMatcher;
use bsnext_guards::MatcherList;
use bsnext_resp::builtin_strings::{BuiltinStringDef, BuiltinStrings};
use bsnext_resp::cache_opts::CacheOpts;
use bsnext_resp::inject_opts::{Injection, InjectionItem};
use bsnext_resp::InjectHandling;
use std::convert::Infallible;
use std::time::Duration;
use tokio::time::sleep;

#[doc = include_str!("./query-params.md")]
#[derive(Debug, serde::Deserialize)]
pub struct DynamicQueryParams {
    /// Allow a request to have a ?bslive.delay.ms=200 style param to simulate a TTFB delay
    #[serde(rename = "bslive.delay.ms")]
    pub delay: Option<u64>,
    /// Control if Browsersync will add cache-busting headers, or not.
    #[serde(rename = "bslive.cache")]
    pub cache: Option<CacheOpts>,
    /// Control if Browsersync will add cache-busting headers, or not.
    #[serde(rename = "bslive.inject")]
    pub inject: Option<InjectParam>,
}

#[derive(Debug, serde::Deserialize)]
#[serde(untagged)]
pub enum InjectParam {
    BuiltinStrings(BuiltinStrings),
    Other(String),
}

#[cfg(test)]
mod test {
    use super::*;
    use http::Uri;
    #[test]
    fn test_deserializing() -> anyhow::Result<()> {
        let input = Uri::from_static(
            "/abc?bslive.delay.ms=2000&bslive.cache=prevent&bslive.inject=bslive:js-connector",
        );
        let Query(query_with_named): Query<DynamicQueryParams> =
            Query::try_from_uri(&input).unwrap();
        insta::assert_debug_snapshot!(query_with_named);

        let input = Uri::from_static("/abc?bslive.inject=false");
        let Query(query_with_bool): Query<DynamicQueryParams> =
            Query::try_from_uri(&input).unwrap();
        insta::assert_debug_snapshot!(query_with_bool);
        Ok(())
    }
}

pub async fn dynamic_query_params_handler(mut req: Request, next: Next) -> impl IntoResponse {
    let Ok(Query(query_params)) = Query::try_from_uri(req.uri()) else {
        let res = next.run(req).await;
        return Ok::<_, Infallible>(res);
    };

    // things to apply *before*
    #[allow(clippy::single_match)]
    match &query_params {
        DynamicQueryParams {
            delay: Some(ms), ..
        } => {
            sleep(Duration::from_millis(*ms)).await;
        }
        _ => {}
    }

    // Other things to apply *before*
    #[allow(clippy::single_match)]
    match &query_params {
        DynamicQueryParams {
            inject: Some(inject_append),
            ..
        } => {
            let uri = req.uri().clone();
            let ex = req.extensions_mut();
            if let Some(inject) = ex.get_mut::<InjectHandling>() {
                tracing::info!(
                    "Adding an item to the injection handling on the fly {} {}",
                    uri.to_string(),
                    uri.path()
                );
                match inject_append {
                    InjectParam::Other(other) if other == "false" => {
                        inject.items = vec![];
                    }
                    InjectParam::BuiltinStrings(str) => inject.items.push(InjectionItem {
                        inner: Injection::BsLive(BuiltinStringDef {
                            name: str.to_owned(),
                        }),
                        only: Some(MatcherList::Item(PathMatcher::pathname(uri.path()))),
                    }),
                    InjectParam::Other(_) => todo!("other?"),
                }
            }
        }
        _ => {}
    }

    let mut res = next.run(req).await;

    // things to apply *after*
    #[allow(clippy::single_match)]
    match query_params {
        DynamicQueryParams {
            cache: Some(cache_opts),
            ..
        } => match cache_opts {
            CacheOpts::Prevent => {
                let headers_to_add = cache_opts.as_headers();
                for (name, value) in headers_to_add {
                    res.headers_mut().insert(name, value);
                }
            }
            CacheOpts::Default => {
                let headers = CacheOpts::Prevent.as_headers();
                for (name, _) in headers {
                    res.headers_mut().remove(name);
                }
            }
        },
        _ => {}
    }

    Ok::<_, Infallible>(res)
}
