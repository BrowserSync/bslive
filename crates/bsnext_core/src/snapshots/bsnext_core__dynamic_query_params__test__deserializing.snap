---
source: crates/bsnext_core/src/dynamic_query_params.rs
expression: query_with_named
---
DynamicQueryParams {
    delay: Some(
        2000,
    ),
    cache: Some(
        Prevent,
    ),
    inject: Some(
        BuiltinStrings(
            JsConnector,
        ),
    ),
}
